### PR TITLE
[DataGrid] Fix column selector crash when hiding columns

### LINKED ISSUE #856: [DataGrid] Column selector crashes with 3 hidden columns
- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

Crash

## Expected Behavior 🤔

No crash

## Steps to Reproduce 🕹

Steps:

1. Open https://material-ui.com/components/data-grid/columns/#column-selector
2. Have 3 columns hidden
3. Capture d’écran 2021-01-16 à 01 19 58

It fails here

Capture d’écran 2021-01-16 à 01 20 46

## Context 🔦

It's a follow-up on #345, @DanailH was leading the effort on it.

### PR DESCRIPTION
**Pull request by DanailH**
Fixes #856

### CODE REVIEW

**Review by oliviertassinari — COMMENTED** [review]
I had a look at the logic, a couple of thought

- The name of `isEqual` is confusing. It's a deep equal, `isDeepEqual` would be clearer.
- I have tried to remove the `if (renderNewColState || newRenderedColState == null) {` branch, and couldn't notice any rendering different. The `updateRenderedCols` method was returning the same order of true/false sequences. If we can remove it, and remove the need for anther state in a ref, that would be even better.
- If we can't I would suggest we isolate the logic more to compute a `const lastVisibleIndexChanged?: boolean` independently.

File: packages/grid/_modules_/grid/hooks/features/virtualization/useVirtualColumns.ts [hunk 1]
```diff
@@ -39,6 +39,8 @@ export const useVirtualColumns = (
   const columnsMeta = useGridSelector(apiRef, columnsMetaSelector);
   const visibleColumnCount = useGridSelector(apiRef, visibleColumnsLengthSelector);
   const visibleColumns = useGridSelector(apiRef, visibleColumnsSelector);
+  const getLastVisibleColIndex = (columns) => (columns.length > 0 ? columns.length - 1 : 0);
```
Comment (dtassone) [inline packages/grid/_modules_/grid/hooks/features/virtualization/useVirtualColumns.ts:42]: Either move this one outside the component or useCallback, because atm you recreate it all the time which makes the further useCallback useless. Also I believe you need to update a dependency array further down

File: packages/grid/_modules_/grid/hooks/features/virtualization/useVirtualColumns.ts [hunk 2]
```diff
@@ -115,17 +117,22 @@ export const useVirtualColumns = (
         lastDisplayedIdx > 0 && (diffLast > tolerance || diffFirst > tolerance);
 
       let newRenderedColState: RenderColumnsProps | null =
-        renderedColRef.current != null ? { ...renderedColRef.current } : null;
-
-      const lastVisibleIndex = visibleColumns.length > 0 ? visibleColumns.length - 1 : 0;
-      if (renderNewColState || newRenderedColState == null) {
+        renderedColRef.current !== null ? { ...renderedColRef.current } : null;
+
+      const newLastVisibleIndex = getLastVisibleColIndex(visibleColumns);
```
Comment (dtassone) [inline packages/grid/_modules_/grid/hooks/features/virtualization/useVirtualColumns.ts:122]: I'm not sure I understand this logic. 
It seems that all we need to do is to generate a new state when we update the visible columns.
Can't we have the same issue on the first col index?

File: packages/grid/_modules_/grid/hooks/features/virtualization/useVirtualColumns.ts (hunk 1 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/features/virtualization/useVirtualColumns.ts]: I cleaned it up, that is no longer needed.

File: packages/grid/_modules_/grid/hooks/features/virtualization/useVirtualColumns.ts (hunk 2 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/features/virtualization/useVirtualColumns.ts]: No there aren't any problems with the first col (or at least I didn't manage to find any). Also, that helper that I created is no longer needed.

File: packages/grid/_modules_/grid/hooks/features/virtualization/useVirtualColumns.ts [hunk 3]
```diff
@@ -39,6 +39,8 @@ export const useVirtualColumns = (
   const columnsMeta = useGridSelector(apiRef, columnsMetaSelector);
   const visibleColumnCount = useGridSelector(apiRef, visibleColumnsLengthSelector);
   const visibleColumns = useGridSelector(apiRef, visibleColumnsSelector);
+  // const getLastVisibleColIndex = (columns) => (columns.length > 0 ? columns.length - 1 : 0);
+  // const lastVisibleIndex = React.useRef<number>(getLastVisibleColIndex(visibleColumns));
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/hooks/features/virtualization/useVirtualColumns.ts:43]:
```suggestion
```

**Review by oliviertassinari — COMMENTED** [review]
I have added a console.log in development to monitor rerenders for scrolling this demo `/components/data-grid/#commercial-version` horizontally in 3 iterations:

**HEAD**

master

**PR**

next

It seems to fly. We can test the UX in production once Netlify has finished building.

**Review by oliviertassinari — COMMENTED** [review]
https://deploy-preview-875--material-ui-x.netlify.app/components/data-grid/#commercial-version seems as smooth as https://master--material-ui-x.netlify.app/components/data-grid/#commercial-version.

**Comment by DanailH** [comment]
> https://deploy-preview-875--material-ui-x.netlify.app/components/data-grid/#commercial-version feels **smoother** than https://master--material-ui-x.netlify.app/components/data-grid/#commercial-version.

**Comment by DanailH** [comment]
Guys is there something else needed for this PR. I was thinking about how to test that corner case but couldn't come up with a good test case (only hiding and showing N columns won't cut it).

**Review by dtassone — APPROVED** [review]

**Review by oliviertassinari — APPROVED** [review]
Yeah, this one is hard to test end-to-end. It's probably the kind of stuff to test with unit tests, like https://devexpress.github.io/devextreme-reactive/ are doing but I'm not sure it's worth it yet (it's a mess for refactoring).
